server-spi/src/main/java/org/keycloak/models/AuthenticationFlowBindings.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.models;
+
+/**
+ * Defines constants for authentication flow bindings.  Strings used for lookup
+ *
+ * @author <a href="mailto:bill@burkecentral.com">Bill Burke</a>
+ * @version $Revision: 1 $
+ */
+public interface AuthenticationFlowBindings {
```

Comment:
Should this be rather turned into enum? Then the JPA `ClientEntity` class would use `@MapKeyEnumerated` rather than `@MapKeyColumn(EnumType.STRING)` and the `AuthenticationFlowResolver` methods could be implemented inside the enum:

```java
public enum AuthenticationFlowBindings {
    BROWSER_BINDING {
        @Override
        public AuthenticationFlowModel resolveFlow(AuthenticationSessionModel authSession) {
            ...
        }
    },
    DIRECT_GRANT_BINDING {
        @Override
        public AuthenticationFlowModel resolveFlow(AuthenticationSessionModel authSession) {
            ...
        }
    };
    public abstract AuthenticationFlowModel resolveFlow(AuthenticationSessionModel authSession);
}
```

That would rid of the String constants and ease adding the new flows in the future since all changes would be needed just in this single class (enum).

Reply:
Protocol plugins are going to have their own flows (i.e. OpenShift needs one), so it needs to be string based. See https://issues.jboss.org/browse/KEYCLOAK-6337

Reply:
I see, `String`s make perfect sense then. Nit: would it be better to use `public static final String` fields and a regular class instead of `interface`?

Reply:
Is that a best practice?  Or is it a Hynek thing?  Over the years I usually try and put constants in an interface for a public API.

Reply:
It is supposed to be best practice (at least according to Joshua Bloch's Effective Java, Item 19 (full explanation in the book, extract in https://github.com/HugoMatilla/Effective-JAVA-Summary#19-use-interfaces-only-to-define-types). But indeed, it may only be Hynek's thing.